sources/api/schnauzer/src/helpers.rs

```diff
@@ -969,77 +983,202 @@ pub fn kube_reserve_cpu(
+
+    // Extract our variables from their serde_json::Value objects
+    let ip_version = ip_version_value
         .as_str()
         .context(error::InvalidTemplateValueSnafu {
             expected: "string",
-            value: hostname_value.to_owned(),
+            value: ip_version_value.to_owned(),
             template: template_name.to_owned(),
         })?;
-    trace!("Hostname string from template: {}", hostname_str);
+    trace!("IP version string from template: {}", ip_version);
 
-    // Attempt to resolve the hostname
-    let hostname_resolveable = match lookup_host(hostname_str) {
-        Ok(ip_list) => {
-            // If the list of IPs is empty or resolves to localhost, consider the hostname
-            // unresolvable
-            let resolves_to_localhost = ip_list
-                .iter()
-                .any(|ip| ip == &IPV4_LOCALHOST || ip == &IPV6_LOCALHOST);
-            if ip_list.is_empty() || resolves_to_localhost {
-                false
-            } else {
-                true
+    let localhost_comparator = match ip_version {
+        "ipv4" => IPV4_LOCALHOST,
+        "ipv6" => IPV6_LOCALHOST,
+        _ => {
+            return Err(error::TemplateHelperError::InvalidTemplateValue {
+                expected: r#"one of ("ipv4", "ipv6")"#,
+                value: ip_version_value.to_owned(),
+                template: template_name.to_owned(),
             }
-        }
-        Err(e) => {
-            trace!("DNS hostname lookup failed: {},", e);
-            false
+            .into());
         }
     };
 
-    // Only write an entry to the template if the hostname is unresolvable
-    if !hostname_resolveable {
-        let ipv4_entry = format!("{} {}", IPV4_LOCALHOST, hostname_str);
-        let ipv6_entry = format!("{} {}", IPV6_LOCALHOST, hostname_str);
-        let entries = format!("{}\n{}", ipv4_entry, ipv6_entry);
-
-        out.write(&entries).context(error::TemplateWriteSnafu {
+    let hostname = hostname_value
+        .as_str()
+        .context(error::InvalidTemplateValueSnafu {
+            expected: "string",
+            value: hostname_value.to_owned(),
             template: template_name.to_owned(),
         })?;
+    trace!("Hostname string from template: {}", hostname);
+
+    let mut results: Vec<String> = vec![];
+
+    let hosts: Option<model::modeled_types::EtcHostsEntries> = (!hosts_value.is_null())
+        .then(|| {
+            serde_json::from_value(hosts_value.clone()).context(
+                error::UnparseableTemplateValueSnafu {
+                    expected: "EtcHostsEntries",
+                    value: hosts_value.to_owned(),
+                    template: template_name.to_owned(),
+                },
+            )
+        })
+        .transpose()?;
+    trace!("Hosts from template: {:?}", hosts);
+
+    // If our hostname isn't resolveable, add it to the alias list.
+    if hostname.len() > 0 && !hostname_resolveable(hostname, hosts.as_ref()) {
```

Comment:
I was trying to think of a way to remove this impurity from this otherwise testable (and complex) function. Alas, I don't see a way to inject this network dependency.

Edit, well... as usual I did think of a way:

`localhost_aliases` could be a thin wrapper for `localhost_aliases_impl`.

`localhost_aliases_impl` takes a functor matching the signature of `hostname_resolvable`, then `localhost_aliases` becomes:

```rust
pub fn localhost_aliases(
    helper: &Helper<'_, '_>,
    h: &Handlebars,
    c: &Context,
    renderctx: &mut RenderContext<'_, '_>,
    out: &mut dyn Output,
) -> Result<(), RenderError> {
    localhost_aliases_impl(helper, h, c, renderctx, out, &hostname_resolvabe)
}
```

Then I think you could write tests for `localhost_aliases_impl` that takes a mock of the `hostname_resolvable` function.

Reply:
I toyed around with implementing something like this, but it ended up making it more challenging to understand and didn't necessarily improve testability, since there are a few tests here already which just use more apparently resolve-able or un-resolve-able hostnames.